src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -264,6 +303,80 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
             .ok_or(StorageError::IdentifierMissing(id))
     }
 
+    /// Persist timestamp bindings updates received from ingestion workers
+    pub(super) async fn persist_timestamp_bindings(
+        &mut self,
+        feedback: &TimestampBindingFeedback<T>,
+    ) {
+        for (id, bindings) in &feedback.bindings {
+            let mut ts_binding_stash = self
+                .storage
+                .stash
+                .collection::<PartitionId, ()>(&id.to_string())
+                .unwrap();
+
+            let collection_state = self.collection_mut(*id).expect("misisng_source_id");
+
+            let upper = ts_binding_stash.upper().unwrap();
+            let since = ts_binding_stash.since().unwrap();
+            let mut seal_frontier = upper.clone();
+
+            // Here we differentialize the bindings we got from workers
+            ts_binding_stash
+                .update_many(bindings.iter().cloned().flat_map(|(pid, ts, offset)| {
+                    let prev_offset = collection_state
+                        .last_reported_ts_bindings
+                        .entry(pid.clone())
+                        .or_default();
+
+                    let ts = ts.try_into().unwrap();
+                    let update = ((pid, ()), ts, offset.offset - prev_offset.offset);
+
+                    if seal_frontier.less_than(&ts) {
+                        seal_frontier = Max1Antichain::Singular(ts);
+                    }
+                    prev_offset.offset = offset.offset;
+                    if upper.less_equal(&ts) {
+                        Some(update)
+                    } else {
+                        None
+                    }
+                }))
+                .unwrap();
+
+            let read_frontier = collection_state.read_capabilities.frontier();
+            assert!(
+                read_frontier.len() <= 1,
+                "unsupported partially ordered time"
+            );
+
+            let mut compaction_frontier: Max1Antichain<i64> = read_frontier
+                .get(0)
+                .map(|x| x.clone().try_into().unwrap())
+                .into();
+            if PartialOrder::less_than(&compaction_frontier, &since) {
+                compaction_frontier = since;
+            }
+            if PartialOrder::less_than(&seal_frontier, &compaction_frontier) {
+                compaction_frontier = seal_frontier.clone();
+            }
+
+            ts_binding_stash.seal(seal_frontier).unwrap();
+            ts_binding_stash.compact(compaction_frontier).unwrap();
+            ts_binding_stash.consolidate().unwrap();
```

Comment:
Oh I meant to ask: do we need to do something smarter than consolidating the timestamp bindings on every update?